tiledb/sm/metadata/metadata.cc

```diff
@@ -0,0 +1,279 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file implements class Metadata.
+ */
+
+#include "tiledb/sm/metadata/metadata.h"
+#include "tiledb/sm/misc/logger.h"
+#include "tiledb/sm/misc/utils.h"
+
+#include <iostream>
+
+namespace tiledb {
+namespace sm {
+
+/* ********************************* */
+/*     CONSTRUCTORS & DESTRUCTORS    */
+/* ********************************* */
+
+Metadata::Metadata() {
+  auto t = utils::time::timestamp_now_ms();
+  timestamp_range_ = std::make_pair(t, t);
+}
+
+Metadata::~Metadata() = default;
+
+/* ********************************* */
+/*                API                */
+/* ********************************* */
+
+void Metadata::clear() {
+  metadata_map_.clear();
+  metadata_index_.clear();
+  loaded_metadata_uris_.clear();
+  timestamp_range_ = std::make_pair(0, 0);
+}
+
+Status Metadata::deserialize(
+    const std::vector<std::shared_ptr<ConstBuffer>>& metadata_buffs) {
+  clear();
+
+  if (metadata_buffs.empty())
+    return Status::Ok();
+
+  uint32_t key_len;
+  std::string key;
+  char del;
+  char value_type;
+  uint32_t value_num;
+  size_t value_len, value_vec_len, offset;
+  std::vector<uint8_t> value, value_vec;
+  for (const auto& buff : metadata_buffs) {
+    // Iterate over all items
+    buff->set_offset(0);
+    while (buff->offset() != buff->size()) {
+      RETURN_NOT_OK(buff->read(&key_len, sizeof(uint32_t)));
+      key.resize(key_len);
+      RETURN_NOT_OK(buff->read((void*)key.data(), key_len));
+      RETURN_NOT_OK(buff->read(&del, sizeof(char)));
+
+      // Handle deletion
+      if (del) {
+        metadata_map_.erase(key);
+        continue;
+      }
+
+      RETURN_NOT_OK(buff->read(&value_type, sizeof(char)));
+      RETURN_NOT_OK(buff->read(&value_num, sizeof(uint32_t)));
+      value_len = value_num * datatype_size(static_cast<Datatype>(value_type));
+      value.resize(value_len);
+      RETURN_NOT_OK(buff->read((void*)value.data(), value_len));
+
+      // Prepare value vector
+      value_vec_len = 2 * sizeof(char) + sizeof(uint32_t) + value_len;
+      value_vec.resize(value_vec_len);
+      std::memcpy(value_vec.data(), &del, sizeof(char));
+      offset = sizeof(char);
+      std::memcpy(value_vec.data() + offset, &value_type, sizeof(char));
+      offset += sizeof(char);
+      std::memcpy(value_vec.data() + offset, &value_num, sizeof(value_num));
+      offset += sizeof(value_num);
+      std::memcpy(value_vec.data() + offset, value.data(), value_len);
+
+      // Insert to metadata
+      metadata_map_.emplace(
+          std::make_pair(std::string(key), std::move(value_vec)));
```

Comment:
You're moving `value_vec` here but then reusing it in the next loop iteration -- this may cause problems.

Reply:
I am resizing it in Line 101. I think that should be ok.

Reply:
I would restructure the code a bit to avoid this. https://en.cppreference.com/w/cpp/utility/move indicates that the state of the object is "unspecified but valid" so it seems like we should just avoid this potentially tricky situation entirely.

Reply:
How about I redifine `value_vec` in Line 101?

Reply:
Yes that seems safer.